opm/simulators/utils/ParallelNLDDPartitioningZoltan.cpp

```diff
@@ -0,0 +1,607 @@
+        }
+
+    private:
+        /// Indices of reachable vertices.
+        std::vector<int> index_{};
+
+        /// Number of reachable vertices.  One more than the maximum value
+        /// in \c index_.
+        int numSeenVertices_{0};
+    };
+
+    /// Unstructured graph of reachable vertices
+    class VertexGraph
+    {
+    public:
+        /// Constructor.
+        ///
+        /// \tparam Edge Type for representing a directed edge between a
+        ///   pair of vertices.  Must support \code std::get<>() \endcode
+        ///   protocol.  Typically a \code std::pair<> \endcode of integral
+        ///   types.
+        ///
+        /// \tparam GlobalCellID Callback type for mapping a vertex ID to a
+        ///   globally unique ID.  Typically the same as
+        ///   \code ParallelNLDDPartitioningZoltan::GlobalCellID \endcode.
+        ///
+        /// \param[in] myRank Current rank in MPI communicator.  Needed by
+        ///   Zoltan.
+        ///
+        /// \param[in] numVertices Maximum number of vertices in graph.
+        ///
+        /// \param[in] edges Edge representation of connectivity graph.
+        ///
+        /// \param[in] vertexId Enumeration of reachable vertices.
+        ///
+        /// \param[in] globalCell Callback for mapping (local) vertex IDs to
+        ///   globally unique vertex IDs.
+        template <typename Edge, typename GlobalCellID>
+        explicit VertexGraph(const int                    myRank,
+                             const std::size_t            numVertices,
+                             const std::vector<Edge>&     edges,
+                             const EnumerateSeenVertices& vertexId,
+                             GlobalCellID&&               globalCell)
+            : myRank_ { myRank }
+        {
+            // Form undirected connectivity graph.
+            for (const auto& [v1, v2] : edges) {
+                this->graph_.addConnection(vertexId[v1], vertexId[v2]);
+                this->graph_.addConnection(vertexId[v2], vertexId[v1]);
+            }
+
+            this->graph_.compress(vertexId.numVertices());
+
+            // Form local-to-global vertex ID mapping for reachable vertices.
+            this->globalCell_.resize(vertexId.numVertices());
+            for (auto vertex = 0*numVertices; vertex < numVertices; ++vertex) {
+                if (const auto localIx = vertexId[vertex]; localIx >= 0) {
+                    this->globalCell_[localIx] = globalCell(vertex);
+                }
+            }
+        }
+
+        /// Retrive my rank in current MPI communicator.
+        ///
+        /// Needed by Zoltan.
+        int myRank() const
+        {
+            return this->myRank_;
+        }
+
+        /// Retrieve number of vertices in connectivity graph.
+        int numVertices() const
+        {
+            return static_cast<int>(this->graph_.numVertices());
+        }
+
+        /// Retrieve globally unqique ID of reachable vertex.
```

Comment:
unique

Reply:
Thanks for noticing. I've pushed an update to fix that.